### PR TITLE
[Exec] Load instead of exec-ing

### PR DESCRIPTION
**Pull request by segiddins**
Proof-of-concept to avoid `exec` where possible, instead just using `load` when we'd be launching the same interpreter anyways. About a .2 second gain in my limited testing. Over a second gain in jruby in my testing, somehow.

Not near ready for merging, but pretty cool for an hour of hacking.

- [x] Tests
- [ ] Centralize error handling
- [ ] Raise exceptions instead of exiting

### CODE REVIEW

File: lib/bundler/cli/exec.rb [hunk 1]
```diff
@@ -37,10 +51,27 @@ def run
...
+
+    def kernel_load(file, *args)
+      args.pop if args.last.is_a?(Hash)
+      ARGV.replace(args)
+      ui = Bundler.ui
+      Bundler.ui = nil
+      require "bundler/setup"
+      Kernel.load(file)
+    rescue SystemExit
+      true
```
Comment (segiddins) [inline lib/bundler/cli/exec.rb]: should just re-raise this

**Comment by segiddins** [comment]
@indirect please eventually talk me out of this...

**Comment by indirect** [comment]
I keep trying to figure out why this would be a bad idea, and I haven't come up with any yet. :P

File: lib/bundler/cli/exec.rb [hunk 2]
```diff
@@ -37,10 +51,26 @@ def run
...
+    rescue Exception => e # rubocop:disable Lint/RescueException
       Bundler.ui = ui
-      Bundler.ui.error "bundler: exec needs a command to run"
-      exit 128
+      Bundler.ui.error "bundler: failed to load command: #{cmd} (#{file})"
+      abort "#{e.class}: #{e.message}\n#{e.backtrace.join("\n  ")}"
+    end
+
+    def ruby_shebang?(file)
+      first_line = File.open(file, "rb", &:readline)
```
Comment (segiddins) [inline lib/bundler/cli/exec.rb]: probably best to only read a fixed number of bytes? an executable might not hit a `line` for a while?

Comment (indirect) [inline lib/bundler/cli/exec.rb]: yeah, but shebangs could have a very long path... maybe 4096 bytes?

Comment (segiddins) [inline lib/bundler/cli/exec.rb]: we're comparing against two known possibilities, can take the max of them

**Comment by segiddins** [comment]
Specs added!

**Comment by indirect** [comment]
@segiddins I'm okay landing this on master now that it has tests... do you think the other changes are required to merge this, or are good changes to make?

**Comment by segiddins** [comment]
Nah, not really. I'd be happy to land this, I guess?

**Comment by indirect** [comment]
:+1: 

@homu r+

**Pull request merged**